// @storybook/angular/dist/ts3.9/demo/button.component.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@angular/core");
let ButtonComponent = class ButtonComponent {
    constructor() {
        this.text = '';
        this.onClick = new core_1.EventEmitter();
    }
};
__decorate([
    core_1.Input(),
    __metadata("design:type", Object)
], ButtonComponent.prototype, "text", void 0);
__decorate([
    core_1.Output(),
    __metadata("design:type", Object)
], ButtonComponent.prototype, "onClick", void 0);
ButtonComponent = __decorate([
    core_1.Component({
        selector: 'storybook-button-component',
        template: ` <button (click)="onClick.emit($event)">{{ text }}</button> `,
        styles: [
            `
      button {
        border: 1px solid #eee;
        border-radius: 3px;
        background-color: #ffffff;
        cursor: pointer;
        font-size: 15px;
        padding: 3px 10px;
        margin: 10px;
      }
    `,
        ],
    })
], ButtonComponent);
exports.default = ButtonComponent;
