// @storybook/angular/dist/ts3.9/client/preview/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.raw = exports.getStorybook = exports.forceReRender = exports.setAddon = exports.clearDecorators = exports.addParameters = exports.addDecorator = exports.configure = exports.storiesOf = void 0;
const core_1 = require("@storybook/core");
require("./globals");
const render_1 = require("./render");
const decorateStory_1 = __importDefault(require("./decorateStory"));
const framework = 'angular';
const api = core_1.start(render_1.renderToDOM, { decorateStory: decorateStory_1.default, render: render_1.render });
exports.storiesOf = (kind, m) => {
    return api.clientApi.storiesOf(kind, m).addParameters({
        framework,
    });
};
exports.configure = (...args) => api.configure(framework, ...args);
exports.addDecorator = api.clientApi
    .addDecorator;
exports.addParameters = api.clientApi
    .addParameters;
exports.clearDecorators = api.clientApi.clearDecorators;
exports.setAddon = api.clientApi.setAddon;
exports.forceReRender = api.forceReRender;
exports.getStorybook = api.clientApi.getStorybook;
exports.raw = api.clientApi.raw;
