// @storybook/angular/dist/ts3.9/client/preview/angular-beta/StorybookModule.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createStorybookModule = exports.getStorybookModuleMetadata = void 0;
const core_1 = require("@angular/core");
const platform_browser_1 = require("@angular/platform-browser");
const ts_dedent_1 = __importDefault(require("ts-dedent"));
const util_deprecate_1 = __importDefault(require("util-deprecate"));
const StorybookProvider_1 = require("./StorybookProvider");
const NgModulesAnalyzer_1 = require("./utils/NgModulesAnalyzer");
const NgComponentAnalyzer_1 = require("./utils/NgComponentAnalyzer");
const StorybookWrapperComponent_1 = require("./StorybookWrapperComponent");
const ComputesTemplateFromComponent_1 = require("./ComputesTemplateFromComponent");
const deprecatedStoryComponentWarning = util_deprecate_1.default(() => { }, ts_dedent_1.default `\`component\` story return value is deprecated, and will be removed in Storybook 7.0.
        Instead, use \`export const default = () => ({ component: AppComponent });\`
        or
        \`\`\`
        export const Primary: Story = () => ({});
        Primary.parameters = { component: AppComponent };
        \`\`\`
        Read more at 
        - https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#deprecated-angular-story-component).
        - https://storybook.js.org/docs/angular/writing-stories/parameters
      `);
exports.getStorybookModuleMetadata = ({ storyFnAngular, component: annotatedComponent, targetSelector, }, storyProps$) => {
    var _a, _b, _c, _d, _e;
    const { component: storyComponent, props, styles, moduleMetadata = {} } = storyFnAngular;
    let { template } = storyFnAngular;
    if (storyComponent) {
        deprecatedStoryComponentWarning();
    }
    const component = storyComponent !== null && storyComponent !== void 0 ? storyComponent : annotatedComponent;
    if (hasNoTemplate(template) && component) {
        template = ComputesTemplateFromComponent_1.computesTemplateFromComponent(component, props, '');
    }
    /**
     * Create a component that wraps generated template and gives it props
     */
    const ComponentToInject = StorybookWrapperComponent_1.createStorybookWrapperComponent(targetSelector, template, component, styles, props);
    // Look recursively (deep) if the component is not already declared by an import module
    const requiresComponentDeclaration = NgComponentAnalyzer_1.isDeclarable(component) &&
        !NgModulesAnalyzer_1.isComponentAlreadyDeclaredInModules(component, moduleMetadata.declarations, moduleMetadata.imports);
    return {
        declarations: [
            ...(requiresComponentDeclaration ? [component] : []),
            ComponentToInject,
            ...((_a = moduleMetadata.declarations) !== null && _a !== void 0 ? _a : []),
        ],
        imports: [platform_browser_1.BrowserModule, ...((_b = moduleMetadata.imports) !== null && _b !== void 0 ? _b : [])],
        providers: [StorybookProvider_1.storyPropsProvider(storyProps$), ...((_c = moduleMetadata.providers) !== null && _c !== void 0 ? _c : [])],
        entryComponents: [...((_d = moduleMetadata.entryComponents) !== null && _d !== void 0 ? _d : [])],
        schemas: [...((_e = moduleMetadata.schemas) !== null && _e !== void 0 ? _e : [])],
        bootstrap: [ComponentToInject],
    };
};
exports.createStorybookModule = (ngModule) => {
    let StorybookModule = class StorybookModule {
    };
    StorybookModule = __decorate([
        core_1.NgModule(ngModule)
    ], StorybookModule);
    return StorybookModule;
};
function hasNoTemplate(template) {
    return template === null || template === undefined;
}
